client/core/core.go

```diff
@@ -5625,6 +5694,116 @@ func (c *Core) authDEX(dc *dexConnection) error {
 	dc.acct.auth(tier, legacyFeePaid)
 
+	// Check active and pending bonds, comparing against result.ActiveBonds. For
+	// pendingBonds, rebroadcast and start waiter to postBond. For
+	// (locally-confirmed) bonds that are not in connectResp.Bonds, postBond.
+
+	// Start by mapping the server-reported bonds:
+	remoteLiveBonds := make(map[string]*msgjson.Bond)
+	remotePendingBonds := make(map[string]*msgjson.Bond)
+	for _, bond := range result.ActiveBonds {
+		if bond.Pending {
+			remotePendingBonds[bondKey(bond.AssetID, bond.CoinID)] = bond
+		} else {
+			remoteLiveBonds[bondKey(bond.AssetID, bond.CoinID)] = bond
+		}
+	}
+
+	// Identify bonds we consider live that are either pending or missing from
+	// server. In either case, do c.monitorBondConfs (will be immediate postBond
+	// and bondConfirmed if at required confirmations).
+	for _, bond := range localActiveBonds {
+		key := bondKey(bond.AssetID, bond.CoinID)
+		symb := dex.BipIDSymbol(bond.AssetID)
+		bondIDStr := coinIDString(bond.AssetID, bond.CoinID)
+
+		_, found := remoteLiveBonds[key]
+		if found {
+			continue // good, it's live server-side too
+		}
+
+		_, found = remotePendingBonds[key]
+		if !found {
+			c.log.Warnf("Locally-active bond %v (%s) not reported by server. Will repost...",
+				bondIDStr, symb) // unexpected, but postbond again
+		}
+
+		bondAsset := bondAssets[symb]
+		if bondAsset == nil {
+			c.log.Warnf("Server no longer supports %v as a bond asset!", symb)
+			continue
+		}
+
+		// Still pending or unknown on server. postBond at required confs.
+		c.log.Infof("Posting locally-confirmed bond %v (%s).", bondIDStr, symb)
+		c.monitorBondConfs(dc, assetBond(bond), bondAsset.Confs)
+		continue
+	}
+
+	// Identify bonds we consider pending that are either live or missing from
+	// server. If live on server, do c.bondConfirmed. If missing, do
+	// c.monitorBondConfs.
+	for _, bond := range localPendingBonds {
+		key := bondKey(bond.AssetID, bond.CoinID)
+		symb := dex.BipIDSymbol(bond.AssetID)
+		bondIDStr := coinIDString(bond.AssetID, bond.CoinID)
+
+		_, found := remoteLiveBonds[key]
+		if found {
+			// It's live server-side. Confirm it locally.
+			c.log.Debugf("Confirming pending bond %v that is confirmed server side", bondIDStr)
+			if err = c.bondConfirmed(dc, bond.AssetID, bond.CoinID, tier /* no change */); err != nil {
+				c.log.Errorf("Unable to confirm bond %s: %v", bondIDStr, err)
+			}
+			continue
+		}
+
+		_, found = remotePendingBonds[key]
+		c.log.Debugf("Starting coin waiter for pending bond %v (%s), (known to server = %v)",
+			bondIDStr, symb, found) // not found is normal
+		bondAsset := bondAssets[symb]
+		if bondAsset == nil {
+			c.log.Warnf("Server no longer supports %v as a bond asset!", symb)
+			continue
+		}
+
+		// Still pending on server. Start waiting for confs.
+		c.log.Debugf("Preparing to post pending bond %v (%s).", bondIDStr, symb)
+		c.monitorBondConfs(dc, assetBond(bond), bondAsset.Confs)
```

Comment:
Tested this case by forcing a server side error on initial bond submission, then on reconnect, retry and success:

```
[INF] CORE: Authenticated connection to 127.0.0.1:17273, acct 9f8d5cbea78335c1c3c789ef1b08c193468adddc6a892e006433bd4ca82e5cb2, 0 active bonds, 0 active orders, 0 active matches, score -1, tier 
[DBG] CORE: Starting coin waiter for pending bond ca9a6c08dc4c973a0524b32a05fc9bf9b28182bed39223508aef190b18d229e3:0 (dcr), (known to server = false)
[DBG] CORE: Preparing to post pending bond ca9a6c08dc4c973a0524b32a05fc9bf9b28182bed39223508aef190b18d229e3:0 (dcr).%!(EXTRA bool=false)
[DBG] CORE: Subscribing to the dcr_btc order book for 127.0.0.1:17273
[DBG] CORE[dcr_btc][book]: Processing 0 cached order notes
[INF] CORE: Bond confirmed ca9a6c08dc4c973a0524b32a05fc9bf9b28182bed39223508aef190b18d229e3:0 (dcr) with expire time of 2022-11-15 10:54:09 -0600 CST
[INF] CORE: Bond ca9a6c08dc4c973a0524b32a05fc9bf9b28182bed39223508aef190b18d229e3:0 (dcr) confirmed.
[INF] CORE: notify: |SUCCESS| (bondpost) BondConfirmed - New tier = 2.
```